[Inliner] Don't count a call penalty for foldable __memcpy_chk and similar

citymarina opened the pull request:
When the size is an appropriate constant, __memcpy_chk will turn into a memcpy that gets folded away by InstCombine. Therefore this patch avoids counting these as calls for purposes of inlining costs.

This is only really relevant on platforms whose headers redirect memcpy to __memcpy_chk (such as Darwin). On platforms that use intrinsics, memcpy and similar functions are already exempt from call penalties.

citymarina committed:
[Inliner] Add tests for memcpy of small constant size. NFC

citymarina commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,48 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
+  // really relevant on platforms whose headers redirect memcpy to __memcpy_chk
+  // (e.g. Mac), as other platforms use memcpy intrinsics, which are already
+  // exempt from the call penalty.
+  if (GetTLI) {
+    auto TLI = GetTLI(*F);
+    LibFunc LF;
+    if (TLI.getLibFunc(*F, LF) && TLI.has(LF)) {
```
In what way do you suggest to exit? The ultimate fallback at the end of the function is the `TTI`, so neither a quick `return true` nor `return false` would be ideal here.

fhahn commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,48 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
+  // really relevant on platforms whose headers redirect memcpy to __memcpy_chk
+  // (e.g. Mac), as other platforms use memcpy intrinsics, which are already
+  // exempt from the call penalty.
+  if (GetTLI) {
+    auto TLI = GetTLI(*F);
+    LibFunc LF;
+    if (TLI.getLibFunc(*F, LF) && TLI.has(LF)) {
```
Perhaps something like below, which would need to duplicate the fallback

```suggestion
TargetLibraryInfo *TLI = GetTLI ? &GetTLI(*F) : nullptr;
LibFunc LF;
if  (!TLI || !TLI->getLibFunc(*F, LF) || !TLI->has(LF) )
    return TI.isLoweredToCall(F);
```

citymarina replied:
I can't say I love the duplication, but OK, done.

nikic commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,48 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
+          auto Len = LenOp->getLimitedValue();
+          auto ObjSize = ObjSizeOp->getLimitedValue();
+          if (ObjSize >= Len &&
+              (Len == 1 || Len == 2 || Len == 4 || Len == 8)) {
```
Do we need the check for these lengths? Even if only these are folded in the middle-end, the back-end will lower fixed-length mallocs that are substantially larger than this to inline load+store expansions as well. It doesn't look like the add a penalty to large llvm.memcpy intrinsics, so we probably shouldn't be doing it here either, if we know we'll fold to llvm.memcpy.

citymarina replied:
Fair point - thanks.

citymarina committed:
[Inliner] Plumb a TargetLibraryInfo through to CallAnalyzer. NFC

citymarina committed:
[Inliner] Don't count a call penalty for foldable __memcpy_chk

When the copy length is known to fit within the object size, calls to
__memcpy_chk will eventually be replaced by inline stores. Therefore this
patch avoids counting these as calls for purposes of inlining costs.

This is only really relevant on platforms whose headers redirect memcpy to
__memcpy_chk (such as Darwin). On platforms that use intrinsics, memcpy
and similar functions are already exempt from call penalties.

citymarina commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,48 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
 }
 
+bool CallAnalyzer::isLoweredToCall(Function *F, CallBase &Call) {
+  // Calls to memcpy with a known constant size 1/2/4/8 should not incur a call
```
Done

citymarina commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,48 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
+        auto LenOp = dyn_cast_or_null<ConstantInt>(Call.getOperand(2));
+        if (!LenOp)
+          LenOp = dyn_cast_or_null<ConstantInt>(
+              SimplifiedValues.lookup(Call.getOperand(2)));
```
Thanks, will look into it as follow-up.

aemerson commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,45 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
+    // Calls to  __memcpy_chk whose length is known to fit within the object
+    // size will eventually be replaced by inline stores. Therefore, these
+    // should not incur a call penalty. This is only really relevant on
+    // platforms whose headers redirect memcpy to __memcpy_chk (e.g. Mac), as
```
Nit: s/Mac/Darwin?

aemerson commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,45 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
+    // platforms whose headers redirect memcpy to __memcpy_chk (e.g. Mac), as
+    // other platforms use memcpy intrinsics, which are already exempt from the
+    // call penalty.
+    {
```
nit: open brace on the case line? Unless this is a clang-format abiding style.

citymarina replied:
In an earlier iteration of this patch, I had to do this to avoid some ridiculous clang-format result, but it seems like things are OK now. Fixed.

aemerson commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2260,6 +2271,45 @@ bool CallAnalyzer::simplifyCallSite(Function *F, CallBase &Call) {
+    // other platforms use memcpy intrinsics, which are already exempt from the
+    // call penalty.
+    {
+      auto LenOp = dyn_cast_or_null<ConstantInt>(Call.getOperand(2));
```
`auto *` for pointers. But also it shouldn't be possible for the call operand to be null so a vanilla `dyn_cast` should suffice?

citymarina replied:
You are right. The `_or_null` was copy-paste residue from `SimplifiedValues.lookup`.

fhahn reviewed (approved):
LGTM, thanks!
